File: yuv_convert/src/main/java/com/zmy/yuv_convert/output/ByteArrayOutput.java
package com.zmy.yuv_convert.output;

import com.zmy.yuv_convert.Format;

public class ByteArrayOutput extends Output<byte[][]> {
    private byte[][] data;

    public ByteArrayOutput(Format format) {
        super(format);
    }


    @Override
    public byte[][] getOutput() {
        ensureDataCapacity();
        int start = 0;
        for (int component = 0; component < format.getComponentCount(); component++) {
            container.position(start);
            int size = format.getComponentCapacity(component, width, height);
            container.limit(start + size);
            start += format.getComponentCapacity(component, width, height);
            container.get(data[component]);
        }
        return data;
    }

    //确保data有足够的容量
    private void ensureDataCapacity() {
        if (data == null || data.length != format.getComponentCount()) {
            data = new byte[format.getComponentCount()][];
        }
        for (int component = 0; component < format.getComponentCount(); component++) {
            if (data[component] == null || data[component].length < format.getComponentCapacity(component, width, height)) {
                data[component] = new byte[format.getComponentCapacity(component, width, height)];
            }
        }
    }
}

File: yuv_convert/src/main/java/com/zmy/yuv_convert/output/ByteArrayOutput.kt
package com.zmy.yuv_convert.output

import com.zmy.yuv_convert.Format

class ByteArrayOutput(format: Format) : Output<Array<ByteArray>>(format) {
    private var ret = mutableListOf<ByteArray>()
    override fun getOutput(): Array<ByteArray> {
        for (i in 0 until format.getComponentCount()) {
            if (i >= ret.size) {
                ret.add(ByteArray(format.getComponentCapacity(i, width, height)))
            } else {
                if (ret[i].size < format.getComponentCapacity(i, width, height)) {
                    ret[i] = ByteArray(format.getComponentCapacity(i, width, height))
                }
            }
        }
        var start = 0
        ret.forEachIndexed { component, buffer ->
            container?.position(start)
            val size = format.getComponentCapacity(component, width, height)
            container?.limit(start + size)
            start += format.getComponentCapacity(component, width, height)
            container?.get(buffer)
        }
        return ret.toTypedArray()
    }
}